apisix/plugins/ai-rag.lua

```diff
@@ -0,0 +1,168 @@
+            },
+            -- change to enum while implementing support for other search services
+            required = { "azure_ai_search" }
+        },
+    },
+    required = { "embeddings_provider", "vector_search_provider" }
+}
+
+local request_schema = {
+    type = "object",
+    properties = {
+        ai_rag = {
+            type = "object",
+            properties = {
+                vector_search = {},
+                embeddings = {},
+            },
+            required = { "vector_search", "embeddings" }
+        }
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1004, -- TODO check with other ai plugins
+    name = "ai-rag",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    -- TODO: check endpoint validity
+    return core.schema.check(schema, conf)
+end
+
+function _M.access(conf, ctx)
+    -- local conf = conf.rag
+    -- if conf then
+    local httpc = http.new()
+    local body_tab = core.request.get_json_request_body_table()
+
+    if not body_tab["ai_rag"] then
+        core.log.error("request body must have \"ai-rag\" field")
+        return 400
+    end
+
+    local embeddings_provider = next(conf.embeddings_provider)
+    local embeddings_provider_conf = conf.embeddings_provider[next(conf.embeddings_provider)]
+    local embeddings_driver = require("apisix.plugins.ai-rag.embeddings." .. embeddings_provider)
+
+    local vector_search_provider = next(conf.vector_search_provider)
+    local vector_search_provider_conf = conf.vector_search_provider[vector_search_provider]
+    local vector_search_driver = require("apisix.plugins.ai-rag.vector-search." .. vector_search_provider)
+
+    local vs_req_schema = vector_search_driver.request_schema
+    local emb_req_schema = embeddings_driver.request_schema
+
+    request_schema.properties.ai_rag.properties.vector_search = vs_req_schema
+    request_schema.properties.ai_rag.properties.embeddings = emb_req_schema
+
+    local ok, err = core.schema.check(request_schema, body_tab)
+    if not ok then
+        core.log.error("request body fails schema check: ", err)
+        return 400
+    end
+
+    local embeddings, err = embeddings_driver.get_embeddings(embeddings_provider_conf, body_tab["ai_rag"].embeddings, httpc)
+    if not embeddings then
+        -- TODO: bring order
+        core.log.error("could not get embeddings: ", err)
+        return 500
+    end
+    core.log.error("dibag err: ", err)
+    core.log.warn("dibag res: ", core.json.encode(embeddings))
+
+    local search_body = body_tab["ai_rag"].vector_search
+    search_body.embeddings = embeddings
+    local res, err = vector_search_driver.search(vector_search_provider_conf, search_body, httpc)
+    if not res then
+        -- TODO: bring order
+        core.log.error("could not get vector_search: ", err)
+        return 500
+    end
+    core.log.error("dibag err: ", err)
+    core.log.warn("dibag res: ", core.json.encode(res, true))
+
+    body_tab["ai_rag"] = nil
```

Comment:
remove `ai_rag` from request body because their purpose is served and also these values will cause failure when proxying requests to LLM.

Reply:
comment the purpose in source code?